qubesfiletrust/qvm_file_trust.py

```diff
@@ -0,0 +1,546 @@
+    """Check for 'user.qubes.untrusted' xattr on the file.
+
+    Expects a readable file.
+    """
+
+    try:
+        file_xattrs = xattr.get_all(path)
+
+    except:
+        error('Unable to read extended attributes of {}'.format(path))
+        safe_chmod(path, orig_perms,
+            'Unable to set original permissions of {}'.format(path))
+        sys.exit(65)
+
+    # Check for our custom qubes attribute
+    untrusted_attribute = (b'user.qubes.untrusted', b'true')
+    if untrusted_attribute in file_xattrs:
+        # We found our attribute
+
+        # Return to original permissions
+        safe_chmod(path, orig_perms,
+            'Unable to set original permissions of {}'.format(path))
+        return True
+
+    # Return to original permissions
+    safe_chmod(path, orig_perms,
+        'Unable to set original permissions of {}'.format(path))
+
+    # We didn't find our attribute
+    return False
+
+def set_visual_attributes(path, attributes_on):
+    """Add visual attributes such as emblems and colors to files/folders"""
+    if attributes_on:
+        try:
+            # Add important nautilius emblem
+            proc = subprocess.Popen(['/usr/bin/gvfs-set-attribute', os.path.realpath(path), '-t', 'stringv',
+                'metadata::emblems', 'important'])
+            subprocess.Popen.wait(proc)
+            os.utime(path, None)
+        except:
+            error('Unable to add emblem \'important\' on {}'.format(path))
+
+    else:
+        try:
+            # Remove important nautilus emblem
+            proc = subprocess.Popen(['/usr/bin/gvfs-set-attribute', os.path.realpath(path), '-t', 'unset',
+                'metadata::emblems'], stdout=subprocess.PIPE)
+            subprocess.Popen.wait(proc)
+            os.utime(path, None)
+        except:
+            error('Unable to remove emblem on {}'.format(path))
+
+def path_is_parent(parent, child):
+    """Check if a child file/path is in a parent folder/path."""
+
+    parent = os.path.abspath(parent)
+    child = os.path.abspath(child)
+
+    return os.path.commonpath([parent]) == os.path.commonpath([parent, child])
+
+def is_untrusted_path(path):
+    """Check to see if the path lies under a path that's considered untrusted
+
+    Files listing untrusted paths lie in /etc/qubes/ and ~/.config/qubes
+    under the name always-open-in-dispvm.list
+    """
+
+    untrusted_paths = retrieve_untrusted_folders()
+
+    # Checks if the path is a child of an untrusted path (might be Linux only)
+    for untrusted_path in untrusted_paths:
+        if path_is_parent(untrusted_path, path):
+            return True
+
+    # Check if untrusted phrase (/etc/qubes/always-open-in-dispvm.phrase) is
+    # present in file path
+    global UNTRUSTED_PHRASE
+    return UNTRUSTED_PHRASE.upper() in path.upper()
+
+def check_file(path, multiple_paths):
+    """Check if the given file is trusted"""
+
+    # Save the original permissions of the file.
+    orig_perms = os.stat(path).st_mode
```

Comment:
As @jpouellet suggested, better don't change permissions, but treat unreadable files as untrusted. But if changing permissions, use `try/finally` block to make sure that it's returned to the original value in all code paths.